src/index.ts

```diff
@@ -0,0 +1,189 @@
+ */
+export interface SnippetOptions {
+  orgId: string;
+  namespace?: string;
+  debug?: boolean;
+  host?: string;
+  script?: string;
+  cookieDomain?: string;
+  recordCrossDomainIFrames?: boolean;
+  recordOnlyThisIFrame?: boolean;
+  devMode?: boolean;
+}
+
+/**
+ * A callback that will be invoked when FullStory has begun a session.
+ *
+ * `sessionUrl` contains the URL to the current session.
+ * `settings`   contains the org settings for the current session.
+ */
+type ReadyCallback = (data: { sessionUrl: string, settings: Readonly<object> }) => void;
+
+declare global {
+  interface Window {
+    _fs_cookie_domain?: string;
+    _fs_debug?: boolean;
+    _fs_dev_mode?: boolean;
+    _fs_host?: string;
+    _fs_initialized?: boolean;
+    _fs_is_outer_script?: boolean;
+    _fs_namespace?: string;
+    _fs_org?: string;
+    _fs_run_in_iframe?: boolean;
+    _fs_script?: string;
+  }
+}
+
+const getFullStory = (): FSApi | undefined => {
+  if (window._fs_namespace) {
+    return window[window._fs_namespace];
+  }
+  return undefined;
+};
+
+const ensureSnippetLoaded = (): FSApi => {
+  const fs = getFullStory();
+  if (!fs) {
+    throw Error(
+      'FullStory is not loaded, please ensure the init function is invoked before calling FullStory API functions'
+    );
+  }
+
+  return fs;
+};
+
+const _init = (inputOptions: SnippetOptions, readyCallback?: ReadyCallback) => {
+  // Make a copy so we can modify `options` if desired.
+  const options = { ...inputOptions };
+  if (getFullStory()) {
+    console.warn('The FullStory snippet has already been defined elsewhere (likely in the <head> element)');
+    return;
+  }
+
+  // see README for details on the recordCrossDomainIFrames option
+  if (options.recordCrossDomainIFrames) {
+    window._fs_run_in_iframe = true;
+  }
+
+  // record the contents of this iFrame when embedded in a parent site
+  if (options.recordOnlyThisIFrame) {
+    window._fs_is_outer_script = true;
+  }
+
+  // Set cookie domain if it was specified.
+  if (options.cookieDomain) {
+    window._fs_cookie_domain = options.cookieDomain;
+  }
+
+  if (options.debug === true) {
+    if (!options.script) {
+      options.script = 'edge.fullstory.com/s/fs-debug.js';
+    } else {
+      console.warn('Ignoring `debug = true` because `script` is set');
+    }
+  }
+
+  initFS(options);
+
+  const fs = getFullStory();
+
+  if (!fs) {
+    console.warn('Trouble with fs');
+    return;
+  }
+
+  if (readyCallback) {
+    fs('observe', { type: 'start', callback: readyCallback });
+  }
+
+  if (options.devMode === true) {
+    const message = 'FullStory was initialized in devMode and will stop recording';
+    fs('trackEvent', {
+      name: 'FullStory Dev Mode',
+      properties: {
+        message_str: message,
+      }
+    });
+    fs('shutdown');
+    window._fs_dev_mode = true;
+    console.warn(message);
+  }
+};
+
+const initOnce = (message) => (inputOptions: SnippetOptions, readyCallback?: ReadyCallback) => {
```

Comment:
Could this return an instance of FSApi?

then the import signature could be `import { init } from '@fulllstory/browser';`

and then client does

```
const FS = init(...);
```

I might be getting out of my depth a little - so feel free to patiently correct my suggestion :)

Reply:
I think we _could_ do this, but we'd need to both have it return the `FullStory` object we're building _and_ continue to export `FullStory`. Basically someone might want to initialize `FullStory` in their `main.js` file and then import `FullStory` in a `component.js` file. 

See this pseudo-coded react sample:
```js
// main.js

import { init } from '@fullstory/browser';

const FS = init(...);

FS('setProperties', { type: 'page', properties: { ... } });

renderApp();

// CartComponent.js

import { FullStory } from '@fullstory/browser';

cosnt CartButton = () => {
  return <button onClick={() => FullStory('trackEvent', { ... })}>Add to cart </button>;
};
```

With that in mind is it too confusing to know that the object returned from `init` is the same as the object you can import in another component? Maybe? Thoughts?

Reply:
Got it. Yeah that's a good callout. `init` should be called, once, as early as possible. `FullStory` can be imported and used anywhere in the component hierarchy. This is the same pattern we institute in `v1` it's just that `init` is a function on `FS`.

I think your original approach is good. Maybe we return an instance of `FSApi` from `inint` in the future if customers request it.